### PR TITLE
Hcal DetIds/Digi DataFrames adjustments to be used with cuda

### PR DESCRIPTION
**Pull request by vkhristenko**
similar to DataFormats/DetId updates for HcalDetid and HcalDigi

- put constexpr 
- move impls to header
- provide the tests
- available functionality allows
  - to get any hcal digi collection (hbhe / hf / ho qie8 ones + qie11/10 ones)
  - ship internals to the device
  - employ DataFrame classes on the gpu.
  - qie8 vs qie10/11 behave quite differently and examples of usage are in the tests

for reference (https://github.com/cms-patatrack/cmssw/pull/58)

__note: this sequence of commits touches DataFormats/Common edm::DataFrame.__ I had to mark most of that class member funcs as constexpr, except for those ctors that take edm::DataFrameContainer, cause then i would need to mark that guy as well

### CODE REVIEW

**Comment by vkhristenko** [comment]
fyi @fwyzard , @makortel

**Comment by Dr15Jones** [comment]
please test

**Comment by Dr15Jones** [comment]
+1
For DataFormats/Common

**Comment by vkhristenko** [comment]
fixed missing default parameters for hcal qie sample

**Comment by makortel** [comment]
@cmsbuild, please test

**Comment by vkhristenko** [comment]
apology for the mess up, fixed missing default parameters in all (_i think at least_) data frames i touched

**Comment by Dr15Jones** [comment]
please test

**Comment by kpedro88** [comment]
CMSSW has generally encouraged complicated functions (more than 1-2 lines) to be defined in .cc files instead of .h files (except for template definitions where this may not work). Will this recommendation be relaxed now for compatibility with different architectures/languages? @fabiocos @davidlange6 @slava77 @Dr15Jones

File: DataFormats/HcalDetId/test/test_hcal_detid.cu [hunk 1]
```diff
@@ -0,0 +1,79 @@
+#include <cuda_runtime.h>
+#include <cuda.h>
+
+#include <iostream>
+#include <assert.h>
```
Comment (kpedro88) [inline DataFormats/HcalDetId/test/test_hcal_detid.cu]: prefer to use `<cassert>` unless CUDA prevents this for some reason

File: DataFormats/HcalDigi/test/test_hcal_digi.cu [hunk 2]
```diff
@@ -0,0 +1,185 @@
+#include <cuda_runtime.h>
+#include <cuda.h>
+
+#include <iostream>
+#include <assert.h>
```
Comment (kpedro88) [inline DataFormats/HcalDigi/test/test_hcal_digi.cu]: also `<cassert>` here

**Comment by vkhristenko** [comment]
@kpedro88 , once the tests finish i will switch the header. thx

**Comment by fwyzard** [comment]
@cmsbuild please test

**Comment by kpedro88** [comment]
`cmsRun: symbol lookup error: /cvmfs/cms-ib.cern.ch/nweek-02526/slc6_amd64_gcc630/cms/cmssw/CMSSW_10_2_X_2018-05-28-2300/lib/slc6_amd64_gcc630/libDataFormatsHGCDigi.so: undefined symbol: _ZN9HcalDetIdC1Ej`
I'm a bit confused because this PR doesn't touch the `HGCDigi` package...

**Comment by vkhristenko** [comment]
it's a symbol for the ctor i believe, although they are left as they were declared/defined

**Comment by kpedro88** [comment]
For some reason `checkdeps` does not checkout the `DataFormats/HGCDigi` package, even though it definitely depends on `HcalDetId` in both the BuildFile and some header files. @smuzaffar do you know why?

**Comment by davidlange6** [comment]
I would guess the issue is that only classes* pieces actually depend on HcalDetId in DataFormats/HGCDigi. Not sure this corner case is well covered by checkdeps

**Comment by smuzaffar** [comment]
this seems like a bug. we should be able to cover the classes.h file dependency too. I will fix checkdeps

**Comment by kpedro88** [comment]
It's actually not just the classes* files, there is a collection defined here: https://github.com/cms-sw/cmssw/blob/master/DataFormats/HGCDigi/interface/HGCDigiCollections.h

**Comment by davidlange6** [comment]
Right - but thats just a header file - it would not by itself trigger a needed recompilation of a package. This header then just gets used in the classes* files. If that header got included into a build product then checkdeps would pick it up (even if it should already being picked up due to classes* usage)

**Comment by smuzaffar** [comment]
Due to the change in build rule (for 10.2.X) to generate filename.cc.o instead of filename.o, the classes.h dependency broke. This should be fixed by https://github.com/cms-sw/cmssw-config/commit/06b346877dd55c9b4e8c881901fafa62e31096cb (in next IB)

**Comment by kpedro88** [comment]
please test

**Comment by civanch** [comment]
+1

**Comment by fwyzard** [comment]
With these changes, GCC 7 complains about
```c++
src/DQM/HcalCommon/src/HashFunctions.cc: In function 'std::__cxx11::string hcaldqm::hashfunctions::name_HFPMiphi(const HcalDetId&)':
src/DQM/HcalCommon/src/HashFunctions.cc:251:15: error: '%d' directive writing between 1 and 4 bytes into a region of size 3 [-Werror=format-overflow=]
   std::string name_HFPMiphi(HcalDetId const& did)
               ^~~~~~~~~~~~~
src/DQM/HcalCommon/src/HashFunctions.cc:251:15: note: directive argument in the range [0, 1023]
src/DQM/HcalCommon/src/HashFunctions.cc:254:11: note: 'sprintf' output between 9 and 12 bytes into a destination of size 10
    sprintf(name, "HF%siphi%d", did.ieta()>0 ? "P" : "M", did.iphi());
    ~~~~~~~^~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~~
```

As `HcalDetId::did` can indeed return a value between 0 and 1023, I would suggest to fix src/DQM/HcalCommon/src/HashFunctions.cc:
```diff
diff --git a/DQM/HcalCommon/src/HashFunctions.cc b/DQM/HcalCommon/src/HashFunctions.cc
index 7135d422589..5f76289a955 100644
--- a/DQM/HcalCommon/src/HashFunctions.cc
+++ b/DQM/HcalCommon/src/HashFunctions.cc
@@ -250,8 +250,8 @@ namespace hcaldqm
 
                std::string name_HFPMiphi(HcalDetId const& did)
                {
-                       char name[10];
-                       sprintf(name, "HF%siphi%d", did.ieta()>0 ? "P" : "M", did.iphi());
+                       char name[12];
+                       snprintf(name, sizeof(name), "HF%ciphi%d", did.ieta()>0 ? 'P' : 'M', did.iphi());
                        return std::string(name);
                }
 
```

**Comment by Dr15Jones** [comment]
@vkhristenko please add the fix proposed by @fwyzard

**Comment by kpedro88** [comment]
@Dr15Jones there is a separate PR: #23471

**Comment by Dr15Jones** [comment]
+1
Then I withdrawl my request.

**Comment by fabiocos** [comment]
+1

the differences in the test histograms look unrelated to this PR

**Pull request merged**